File: Padrões Gof/Estruturais/Decorator/Decorator01.js
// Componente
class ComponenteNotificador{
  enviar(mensagem){
    console.log(`Enviando Mensagem:${mensagem}`);
  }
}

// Decorator base
class DecoratorNotificador{
  constructor(componente){
    this.componente = componente;
  }

enviar(mensagem){
  this.componente.enviar(mensagem);
  }
}
// Decorator Concreto 1
class DecoratorNotificadorSMS extends DecoratorNotificador{
  enviar(mensagem){
    super.enviar(mensagem);
    console.log(`Enviar mensagem por SMS:${mensagem}`);
  }
}

// Decorator Concreto 2
class DecoratorNotificadorFacebook extends DecoratorNotificador{
  enviar(mensagem){
      super.enviar(mensagem);
      console.log(`Enviar mensagem pelo Facebook: ${mensagem}`);
    }
  }
// Cliente #############
const notificador = new ComponenteNotificador();
const notificadorComSMS = new DecoratorNotificadorSMS(notificador);
const notificadorComFacebook = new DecoratorNotificadorFacebook(notificadorComSMS)

notificadorComFacebook.enviar("Olá Mundo...");

File: Padrões Gof/Estruturais/Decorator/Decorator01.py
# Componente
class ComponenteNotificador:
    def enviar(self, mensagem):
        print(f"Enviando mensagem: {mensagem}")



# Decorador Base
class DecoradorNotificador:
    def __init__(self, componente):
        self.componente = componente


    def enviar(self, mensagem):
        self.componente.enviar(mensagem)



# Decorador Concreto 1
class DecoradorNotificadorSMS(DecoradorNotificador):
    def enviar(self, mensagem):
        super().enviar(mensagem)
        print(f"Enviando mensagem por SMS: {mensagem}")



# Decorador Concreto 2
class DecoradorNotificadorFacebook(DecoradorNotificador):
    def enviar(self, mensagem):
        super().enviar(mensagem)
        print(f"Enviando mensagem pelo Facebook: {mensagem}")



# CLIENTE ####################
notificador = ComponenteNotificador()
notificadorComSMS = DecoradorNotificadorSMS(notificador)
notificadorComFacebook = DecoradorNotificadorFacebook(notificadorComSMS)

notificadorComFacebook.enviar("Ola Mundo...")

